src/r1cs/sparse.rs

```diff
@@ -0,0 +1,249 @@
+#[abomonation_bounds(where <F as PrimeField>::Repr: Abomonation)]
+pub struct SparseMatrix<F: PrimeField> {
+  #[abomonate_with(Vec<F::Repr>)]
+  /// all non-zero values in the matrix
+  pub data: Vec<F>,
+  /// column indices
+  pub indices: Vec<usize>,
+  /// row information
+  pub indptr: Vec<usize>,
+  /// number of columns
+  pub cols: usize,
+}
+
+impl<F: PrimeField> SparseMatrix<F> {
+  /// 0x0 empty matrix
+  pub fn empty() -> Self {
+    SparseMatrix {
+      data: vec![],
+      indices: vec![],
+      indptr: vec![0],
+      cols: 0,
+    }
+  }
+
+  /// construct from Vec<usize(row), usize(col), F>
+  pub fn new(matrix: &[(usize, usize, F)], rows: usize, cols: usize) -> Self {
+    let mut new_matrix = vec![vec![]; rows];
+    for (row, col, val) in matrix {
+      new_matrix[*row].push((*col, *val));
+    }
+
+    for col in new_matrix.iter_mut() {
+      col.sort_by(|(x, _), (y, _)| x.cmp(y));
+    }
+
+    let mut indptr = vec![0; rows + 1];
+    for (i, col) in new_matrix.iter().enumerate() {
+      indptr[i + 1] = indptr[i] + col.len();
+    }
+
+    let mut indices = vec![];
+    let mut data = vec![];
+    for col in new_matrix {
+      let (idx, val): (Vec<_>, Vec<_>) = col.into_iter().unzip();
+      indices.extend(idx);
+      data.extend(val);
+    }
+
+    SparseMatrix {
+      data,
+      indices,
+      indptr,
+      cols,
+    }
+  }
+
+  /// retrieves the data for row slice [i..j] from ptrs
+  pub fn get_row_unchecked(&self, ptrs: &[usize]) -> impl Iterator<Item = (&F, &usize)> {
+    self.data[ptrs[0]..ptrs[1]]
+      .iter()
+      .zip(&self.indices[ptrs[0]..ptrs[1]])
+  }
+
+  /// multiply by dense vector; uses rayon/gpu
+  pub fn multiply_vec(&self, vector: &[F]) -> Vec<F> {
+    assert_eq!(self.cols, vector.len(), "invalid shape");
+
+    self.multiply_vec_unchecked(vector)
+  }
+
+  /// multiply by dense vector; uses rayon/gpu
+  /// this does not check the shape of the matrix/vector
+  #[tracing::instrument(skip_all, name = "SparseMatrix::multiply_vec_unchecked")]
+  pub fn multiply_vec_unchecked(&self, vector: &[F]) -> Vec<F> {
+    self
+      .indptr
+      .par_windows(2)
+      .map(|ptrs| {
+        self
+          .get_row_unchecked(ptrs)
+          .map(|(val, col_idx)| *val * vector[*col_idx])
+          .sum()
+      })
+      .collect()
+  }
+
+  /// number of non-zero entries
+  pub fn len(&self) -> usize {
+    *self.indptr.last().unwrap()
+  }
+
+  /// empty matrix
+  pub fn is_empty(&self) -> bool {
+    self.len() == 0
+  }
+
+  /// returns a custom iterator
+  pub fn iter(&self) -> Iter<'_, F> {
+    let mut row = 0;
+    while self.indptr[row + 1] == 0 {
+      row += 1;
+    }
+    Iter {
+      matrix: self,
+      row,
+      i: 0,
+      nnz: *self.indptr.last().unwrap(),
+    }
+  }
+
+  /// pad to specific dimensions, which must be bigger
+  pub fn pad(&mut self, rows: usize, num_vars_padded: usize, num_vars: usize) {
```

Comment:
Could you add a unit test for the padding please?

Reply:
The change in logic is completely fine, but I expect a unit test to show what's expected of this function. It's doubly needed if the internal logic is *not* what one would usually infer from the function's name.

Reply:
Ok, I see what you mean. Would it make more sense to put the test in the `R1CSShape` module, though, since that is the central object being modified?